pallets/oracle-feed/src/lib.rs

```diff
@@ -0,0 +1,149 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version (see http://www.gnu.org/licenses).
+
+// Centrifuge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+//! Oracle pallet to feed values.
+//! Feeding is permissionless given an initial fee for each key.
+//! This pallet do not aggregate/validate any, it just store them by account as
+//! they come. It's expected other pallet read this storage and
+//! aggregate/validate the values.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+
+pub mod weights;
+
+pub use pallet::*;
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cfg_traits::{fees::PayFee, ValueProvider};
+	use frame_support::{pallet_prelude::*, traits::Time};
+	use frame_system::pallet_prelude::*;
+
+	use crate::weights::WeightInfo;
+
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	type MomentOf<T> = <<T as Config>::Time as Time>::Moment;
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Identify an oracle value
+		type OracleKey: Parameter + Member + Copy + MaxEncodedLen;
+
+		/// Represent an oracle value
+		type OracleValue: Parameter + Member + Copy + MaxEncodedLen + Default;
+
+		/// A way to obtain the current time
+		type Time: Time;
+
+		/// Fee for the first time a feeder feeds a value
+		type FirstValuePayFee: PayFee<Self::AccountId>;
+
+		/// The weight information for this pallet extrinsics.
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Store all oracle values indexed by feeder
+	#[pallet::storage]
+	pub(crate) type FedValues<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId,
+		Blake2_128Concat,
+		T::OracleKey,
+		(T::OracleValue, MomentOf<T>),
+	>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		Fed {
+			account_id: T::AccountId,
+			key: T::OracleKey,
+			value: T::OracleValue,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Permissionles call to feed an oracle key from a source with value.
+		/// The first time the value is set, an extra fee is required for the
+		/// feeder.
+		#[pallet::weight(T::WeightInfo::feed_first())]
+		#[pallet::call_index(0)]
+		pub fn feed(
+			origin: OriginFor<T>,
+			key: T::OracleKey,
+			value: T::OracleValue,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
```

Comment:
We should just ensure it is not unsigned. All other origins are fine.

Reply:
I would say there is no need to pay fees for the case of root, right?

Reply:
Actually, I have no way to do it, since fees are changed to an `AccountId`

Reply:
That is fine, only `Origin::Signed` should pay a fee.